plutus-tx-plugin/test/Budget/9.6/andExpensive.uplc.golden

```diff
@@ -2,19 +2,15 @@ program
   1.1.0
   ((\s ->
       s s)
-     (\s x ->
-        (\and ->
-           force
-             (case
-                x
-                [ (delay (constr 0 []))
-                , (\x xs ->
-                     delay
-                       (force
-                          (case
-                             x
-                             [(delay (and xs)), (delay (constr 1 []))]))) ]))
-          (s s))
+     (\s ds ->
+        force
+          (case
+             ds
+             [ (delay (constr 0 []))
+             , (\x xs ->
+                  delay
+                    (force
+                       (case x [(delay (s s xs)), (delay (constr 1 []))]))) ]))
```

Comment:
> Why does the optimizer not do that? Perhaps 'cause it's not semantics preserving, for all the optimizer knows evaluating s s (I just realized I should've named it differently...) may be impure.

That's exactly why.

What I'm wondering is how the cost improvement can be so significant. I'd expect a very tiny improvement, and if `and` is used more than once, I'd expect a regression :confused:

Reply:
> I'd expect a very tiny improvement

It may simply enable us to do more optimizations, I'll take a closer look. But in any case not saving something into a name pointlessly and then needing to do a lookup certainly should indeed help.

> if `and` is used more than once, I'd expect a regression 😕

No, `and` is computed at every recursive step: the `\s -> s s` thing, i.e. the driver of recursion, is outside of it.

Reply:
> It may simply enable us to do more optimizations

But this is the final UPLC, after all optimizations. I struggle to see how the cost difference can be so big - I must be missing something.

> No, and is computed at every recursive step: the \s -> s s thing, i.e. the driver of recursion, is outside of it.

What I'm saying is that if `and` occurs twice, then inlining it causes `s s` to be evaluated twice at each recursive step, which should be worse. I'm surprised that there isn't a single test case that became worse.

Reply:
> I now wonder if we can make it even more efficient by somehow caching `s s` once and for all.

I've looked at it as hard as I could and I failed to find any opportunity there, it probably doesn't exist.